Update Index Scan Costing for index columns not present in predicate and vice versa

24nishant opened the pull request:
Through  this PR the costing for Index , Index Only, Dynamic Index and Dynamic Index Only scans is updated.  

_**Terminology**_
`If,`
Table  columns       : a, b, c, d, e
Index columns        : a, b, c
Predicate columns : c, d, e
`Then,`
**Used Index columns**                  : c       (The columns which are present in both index and predicate.)
**Unused Index columns**              : a, b  (The columns which are present in the index but not in the predicate.)
**Unindexed Predicate columns**  : d, e  (The columns which are present in the predicate but not in the index)
 


### **MOTIVATION**
----------------------------------------------------
ORCA produces same cost for Index scan in following scenarios
```
create table foo (a int , b int, c int) distributed by (a);
insert into foo select i,i,i from generate_series(1,5)i;
insert into foo select 1,1,1*i/i from generate_series(1,5)i;
analyze foo;
```
**1. Unindexed Predicate columns** 
For the same index, different number of predicate columns are unindexed in the query.
Eg: For Index - idx_foo_a ; predicate - a vs predicate - a,b vs predicate - a,b,c

```
create index idx_foo_a ON foo USING btree(a);
explain analyze select * from foo where a=1;
explain analyze select * from foo where a=1 and b=1;
explain analyze select * from foo where a=1 and b=1 and c=1;
```

**2. Unused Index columns** 
For the same index, the query is conditioned on different index columns.
Eg: For Index - idx_foo_ab ; predicate - a vs predicate - b vs predicate - a,b
```
create index idx_foo_ab ON foo USING btree(a,b);
analyze foo;
explain analyze select * from foo where a=1;
explain analyze select * from foo where b=1;
```


**3. Index columns ordering**
Given two indexes, idx_foo_ab and idx_foo_ba, if the predicate is on 'a',
ORCA produces same scan cost for both indexes.
The same is applicate if the predicate is on 'b'.
```
create index idx_foo_ab ON foo using btree(a,b);
create index idx_foo_ba ON foo using btree(b,a);
analyze foo;
-- index - idx_ab should be picked
explain analyze select * from foo where a=1;
-- index - idx_ba should be picked
explain analyze select * from foo where b=1;
```


**This PR introduces the following corrections** :

1.  Consider effect of 'unused Index column(s)' in predicate for costing.

            Scenarios: 
           Index :  idx_foo_ab ; predicate is on - 'a' ; Unused index column in predicate : 'b'
           Index :  idx_foo_ab ; predicate is on - 'b' ; Unused index column in predicate : 'a'

2.  Consider effect of UnIndexed predicate column(s) in index for costing.

         Scenarios: 
           Index :  idx_foo_a ; predicate is on - a,b ;    Unindexed predicate column : 'b'
           Index :  idx_foo_a ; predicate is on - a,b,c ; Unindexed predicate column : 'b','c'


New costing equations are introduced to include the effect of above cases in the cost model:

```
-   Unused Index Cost  = Unused index weight with NDV * Cost Conversion factor
Unused index weight with NDV =  Unused index weight  * (NDV of unused column/total rows in table)

-   Unindexed Pred Cost = Unindexed Pred Count * Cost Conversion factor
```


### **IMPLEMENTATION DETAILS**
----------------------------------------------------
Cost is computed for both unused index columns, and unindexed predicate columns.

The cost computation is based on following:
1. For unindexed predicate columns, we account for the number of unindexed predicate columns.

2. For unused index columns, weight is computed based on the column's position in the index.

3. Ratio of (NDV of unused index column )/(Total no of tuples in table), is used to
  modify the impact of unused index column.
  Eg: If the ratio=1, it means all the rows are unique. Therefore,
  full cost penalty should be considered if this column is unused because this
  column was critical for search.

4. The weight/count is multiplied by a cost conversion factor of '1.0e-04'. The value is chosen to differentiate
the cost of different index scan alternatives, but not result in changes of table scan operators, such as Bitmap
scans & Sequential scans. 

(Both the bitmap scan and sequential scan has a "base cost" mapped to `DInitScanFacorVal` (431.0). Index
scan doesn't have a "base cost".)



### **TESTING**
----------------------------------------------------
Please check the attached sheet, where all the scenarios for Index, Index Only
Dynamic Index and Dynamic Index Only scans are covered with costing details.
A comparision is presented, with the existing costing vs the updated costing.
For each of the scenarios, detailed computation is done as implemeted in the
code for the ease of verification.
[ResultsForVerification.xlsx](https://docs.google.com/spreadsheets/d/1RwPQFjDP5kl58WCmJecK8ybua9rcg1Eg/edit?usp=sharing&ouid=105599807536520562474&rtpof=true&sd=true)


### **RESULTS**
----------------------------------------------------
**1. Unindexed Predicate columns** 

<img width="473" alt="image" src="https://github.com/greenplum-db/gpdb/assets/99574850/4825c216-7820-46f3-8d48-a7f28a96bdf6">


**2. Unused Index columns** 

<img width="470" alt="image" src="https://github.com/greenplum-db/gpdb/assets/99574850/fa872430-5ff2-4f65-8af9-f2403460d04f">

**3. Index columns ordering**

<img width="470" alt="image" src="https://github.com/greenplum-db/gpdb/assets/99574850/16e2cd90-d344-4330-a660-9ef7a5c6f45e">

### **Other Reference**
----------------------------------------------------
I have created [list](https://docs.google.com/spreadsheets/d/1rj_Vy0-jJCLSbuc_FKU6FAs1Q4KjJ4QB/edit?usp=share_link&ouid=105599807536520562474&rtpof=true&sd=true) of all the MDP changes in the PR for ease of review.

Out of 178 MDPs only 44 have costing changes and 2 in these have plan change (on account of another index selection). Rest all the other MDPs have either addition /deletion of flags like PartIndexId, FuncVariadic etc

chrishajas commented:
I'm surprised to MDP changes even in tests that don't involve indexes. Also, can you squash the commits to make reviewing easier--A commit (or multiple commits) for code changes and a separate commit for MDP changes.  You can squash them all together once you merge, but having just a couple separate commits makes it easier to review/load the page.

24nishant commented:
> ou squash the commits to make review



> I'm surprised to MDP changes even in tests that don't involve indexes. Also, can you squash the commits to make reviewing easier--A commit (or multiple commits) for code changes and a separate commit for MDP changes. You can squash them all together once you merge, but having just a couple separate commits makes it easier to review/load the page.

Thank you @chrishajas  for the feedback. Would request you to please take a look at the following reasons, due to which I created these commits. I created different commits in order to highlight critical changes in the code. 
I am open to squash them, if you still feel so.

**Commit 1** : In this, all the basic functionality for Index, Index Only and Dynamic Index scan is added

**Commit 2** : Here, I computed the NDV for missing index columns. Also one critical change was made in 
                           ComputeTableStats(), where generation of base table stats was allowed. This changes caused some 
                           MDPs to fails. To highlight this change, I kept it separate.

**Commit 3** :  In this commit, all the code changes for cost conversion factor are compiled to highlight the carefully       chosen cost  conversion factor.

**Commit 4** : Formatting Changes 

**Commit 5** : Updated code for Dynamic Index only scans. 

**Commit 6** : Added ICW tests 

**Commit 7** : Added new MDPs to test the code

**Commit 8** : Compiled all the MDPs where majorly costing changed, and were fixed by the 'fix_mdp.py' utility 

**Commit 9** : This commit has details of MDPs which raised Exceptions, after computation of stats for base table was 
allowed. Since there were more than 400 MDPs changes, i made a separate commit so as to highlight them,

**Commit 10** : Here I addressed pipeline failures.

dgkimura reviewed (commented):
Bold change.  Nice work.  Left a few comments after reading tests and skimming code.  Yet to look more closely at code.

Any explain pipeline changes?

dgkimura commented on src/backend/gporca/libgpopt/src/operators/CPhysicalScan.cpp:
```diff
@@ -256,6 +256,18 @@ CPhysicalScan::ComputeTableStats(CMemoryPool *mp)
 	CColRefSet *pcrsHist = GPOS_NEW(mp) CColRefSet(mp);
 	CColRefSet *pcrsWidth = GPOS_NEW(mp) CColRefSet(mp, m_pdrgpcrOutput);
 
+	// All table columns, excepts system columns are included
+	// for generation of Histogram.
+	CColRefSetIter crsiHist(*pcrsWidth);
+	while (crsiHist.Advance())
+	{
+		CColRef *colref = crsiHist.Pcr();
+		if (!colref->IsSystemCol())
+		{
+			pcrsHist->Include(colref);
+		}
+	}
```
It may be possible to leverage `CColRef::GetUsage()` here to prevent generating unnecessary histograms.

24nishant replied:
Thanks @dgkimura , I have used `CColRef::GetUsage()` here, will push the changes in the next commit.

dgkimura commented on src/test/regress/sql/gp_index_costing.sql:
```diff
@@ -0,0 +1,115 @@
+--
+-- Test Cases to check Index scan, Index Only scan and Dynamic Index scan costing
+--
+
+
+---------------------------------------------------------------------
+-- Scenario1: Verify impact of missing index column in predicate,
+-- which helps in selecting best index in a multiple index scenario.
+---------------------------------------------------------------------
+
+    -- Case A : Index - idx_ab, Columns used in predicate : a
+    -- Case B : Index - idx_ab, Columns used in predicate : b
+    -- In the above two cases, the index scan using idx_ab should be higher
+    -- in CASE B, compared for CASE A, as the in Case-B, index column
+    -- of higher significance is missing  i.e. col 'a'
```
I couldn't think of any good way to benchmark this currently (w/o plan hints).  A couple of hack alternatives:
  1) create multiple tables with the same dataset and different index types, then compare queries' performance.
  2) create single table with multiple indexes and set `optimizer_plan_id` to benchmark the different indexes.
  
The 2nd approach might not be as bad as it sounds, because I think the number of alternatives should be small.  I'm curious if anybody has thoughts/opinions.  Maybe the answer is to do nothing until we have plan hints and I think I could be okay with that for now.

24nishant replied:
@dgkimura , Thank you for the feedback. 
Since it is difficult to show effect of costing in ICW test case, I prepared a separate excel sheet covering different costing scenarios.  I would request to you please check the excel sheet attached in the description [ResultsForVerification.xlsx](https://github.com/greenplum-db/gpdb/files/12359112/ResultsForVerification.xlsx)

24nishant commented on src/test/regress/expected/qp_indexscan_optimizer.out:
```diff
@@ -16,7 +16,7 @@ explain (costs off) select a from test_index_with_orderby_limit order by a limit
          Merge Key: a
          ->  Limit
-               ->  Index Scan using index_a on test_index_with_orderby_limit
+               ->  Index Only Scan using index_a on test_index_with_orderby_limit
```
Thanks for pointing this out, I have corrected this, will push in the next commit. 

- I checked in the code, after the updated costing logic, for all of these queries, the cost of Index scan was higher than Index only scan. It was due to additional cost on account of 'missing predicate column'. 

- I am retrieving the 'Number of missing predicate columns in index' from [pdrgpexprResidual](https://github.com/greenplum-db/gpdb/blob/b3f99d89d10ed7a9af524cee567774fd85593e11/src/backend/gporca/libgpopt/src/xforms/CXformUtils.cpp#L2474), an output from CPredicateUtils::ExtractIndexPredicates(). 

- My understanding was that, the size of 'pdrgpexprResidual' will be ZERO, if there are no predicates in the query, but in a case, where we have 'only order by clause' , its size is not zero. For eg: `explain (costs off) select a from test_index_with_orderby_limit order by a limit 10;`. This non-zero value of 'missing predicate column in the Index', increased the cost for Index scans. 

- Now I have changed it to read size of pdrgpexprResidual, only if there is any predicate on the index column.

24nishant commented on src/test/regress/sql/gp_index_costing.sql:
```diff
@@ -0,0 +1,115 @@
+--
+-- Test Cases to check Index scan, Index Only scan and Dynamic Index scan costing
```
I have updated the comment for Dynamic Index only scans. 
I will check and update, if want to implement these improvements for bitmap scan now.

24nishant commented on src/test/regress/sql/gp_index_costing.sql:
```diff
@@ -0,0 +1,115 @@
+CREATE INDEX idx_foo_ba ON foo USING btree(b,a);
+analyze foo;
+    -- index idx_foo_ab should be selected
+explain select * from foo where a=1;
```
I have added more tests in the excel sheet attached in the description. please check sheet 'other tests' in [ResultsForVerification.xlsx](https://github.com/greenplum-db/gpdb/files/12359112/ResultsForVerification.xlsx)

THANATOSLAVA commented:
> **Scenarios where updated costing logic helps**

I suggest moving the above section to the top as the **motivation** for this work. Before learning about the implementation, I would like to know why the work is needed. Also, can we include the behavior before the costing update for all the three scenarios? This helps highlight the impact of the work.

THANATOSLAVA commented:
> Cost conversion factor is selected as '1.0e-04' because using the additional cost factor,
primary aim is to differentiate between 'plans' with same cost. Thus, a small value
is sufficient to differentiate between them.

1. Do you mean the primary goal is to differentiate index scan plans with the same cost?
 
2. If so, how likely is for index scan that used to cost less than other scan types (sequential, bitmap) to cost more with this additional cost?

THANATOSLAVA commented:
The following two examples could use some rewording. 

> Differentiate in index scan cost if for the same index, the query is on different
index columns, earlier this cost was same.
Eg: For Index - idx_foo_ab ; predicate - a vs predicate - b vs predicate - a,b

And just to be sure, this example illustrates the missing pred cost?

> Differentiate in index scan cost if for the same index, different predicate
columns are missing in the queries, earlier this cost was same.
Eg: For Index - idx_foo_a ; predicate - a vs predicate - a,b vs predicate - a,b,c

and this example illustrates the missing index cost?

And before this PR, if there's any overlap between the index and predicate columns, we cost the index scan the same amount?

THANATOSLAVA commented:
> Consider effect of missing Index column(s) in predicate for costing.
Consider effect of missing predicate column(s) in table index for costing.

We could use an example here...

Say table foo is indexed on `a,b`. The query predicates are on columns `b`, `c`. Column `a` is a missing index column, and column `d` is a missing predicate column.

Frankly, I find the wording of "missing index/predicate column" really confusing. Cause one can argue both ways:

1. missing (index column) --> an index column that's missing from the predicates
2. (missing index) column --> a predicate column that's not indexed

I propose we rephrase "missing index/predicate column" as:

missing index column --> unused index column
missing predicate column -> unindexed column

24nishant commented:
> Bold change. Nice work. Left a few comments after reading tests and skimming code. Yet to look more closely at code.
> 
> Any explain pipeline changes?

Thank you @dgkimura.
As we checked and discussed offline

1. Plan diffs were found in the 'Explain pipeline' run for this PR branch.
2. These were due to the added logic of ' Enabling computation of statistics for columns used in the query for the table ([here](https://github.com/24nishant/gpdb/blob/0530da6242a3ab2d3c73c7e8e6290f9d2c3ef030/src/backend/gporca/libgpopt/src/operators/CPhysicalScan.cpp#L260))' 
3. However, there are no 'diffs' in the 'Explain pipeline' due to change in costing.
4. This was ascertained, by creating a new branch (out of this PR branch) & removing the above described logic for statistics computation. On running, 'explain pipeline' for this new branch, no plan diffs were found.

24nishant commented on src/test/regress/sql/gp_index_costing.sql:
```diff
@@ -0,0 +1,115 @@
+insert into foo select 1,1,1*i/i from generate_series(1,5)i;
+
+-- 1.1 Test case for index scans
+CREATE INDEX idx_foo_ab ON foo USING btree(a,b);
```
I tested for Hash, GIST & GIN based index. I have posted all the results in 'DIFFERENT INDEX TYPES' sheet of  [ResultsForVerification.xlsx](https://docs.google.com/spreadsheets/d/1RwPQFjDP5kl58WCmJecK8ybua9rcg1Eg/edit?usp=sharing&ouid=105599807536520562474&rtpof=true&sd=true)

Overview:

1. **Hash index** supports, index creation on single column only.
Thus, a scenarios of 'missing/unused  index column' in the predicate is not possible.
Only missing/Unused predicate column in the index scenario is possible.The performance in this case is simmilar to  B-Tree index case. An additional cost is included for every missing/unused predicate in the index.

2. **GIST Index**

The additional cost logic worked for this index also.  Tested it for both the scenarios of Missing/unused index and/or predicate columns

3. **GIN & BRIN  index**
For the queries I ran for testing GIN & BRIN indexes,  by default, sequential scan is selected , if sequential scan is disabled then Bit map heap scan is selected. On disabling Bitmap scan, the query falls back to planner.

Please share your feedback, if I need to look in some other test cases.

THANATOSLAVA commented on src/backend/gporca/libgpdbcost/src/CCostModelGPDB.cpp:
```diff
@@ -1662,11 +1801,29 @@ CCostModelGPDB::CostIndexScan(CMemoryPool *,  // mp
 	return CCost(pci->NumRebinds() *
-				 (dRowsIndex * dCostPerIndexRow + dIndexScanTupRandomFactor));
+				 (dRowsIndex * dCostPerIndexRow + dIndexScanTupRandomFactor +
+				  dTotalAdditionalCost));
```
We can remove `Total` from all the variable names. If it's `Cost`, it's the total cost. Otherwise it's either `CostUnit` or `CostPer`.

Should probably also not call it "additional" cost. "Additional" doesn't say what it is. We can just add both the `Pred` and `Index` components to the cost equation.

THANATOSLAVA commented:
I am not persuaded that `m_ulResidualPredicateSize` belongs to the `IndexScan/Get` classes. I see the convenience obtaining its value in `CXformUtils`, where we derive the residual predicates expression array. But we should only add a class member variable if the variable speaks of what this class is, or, if we want the variable to be easily accessible. That's not the case with `m_ulResidualPredicateSize`. When we think of `IndexScan/Get`, I don't see the number of the predicates that we have to filter again on each row fetched a key signature of the class. And the only occasion that we access the variable is in the cost model. In my opinion, it makes a lot more sense to derive the number of residual predicates in `CCostModelGPDB`. Actually we can even make `ComputeAdditionalMIssingIndexWeight` dual-purposed, where we have access to both index and predicate columns. The first round we locate the indices of the index columns unused by predicates. The second round we locate the count of unindexed predicate columns.

24nishant commented:
> > **Scenarios where updated costing logic helps**
> 
> I suggest moving the above section to the top as the **motivation** for this work. Before learning about the implementation, I would like to know why the work is needed. Also, can we include the behavior before the costing update for all the three scenarios? This helps highlight the impact of the work.

Thank you @THANATOSLAVA for the feedback, I have updated the description.

24nishant commented:
> > Cost conversion factor is selected as '1.0e-04' because using the additional cost factor,
> > primary aim is to differentiate between 'plans' with same cost. Thus, a small value
> > is sufficient to differentiate between them.
> 
> 1. Do you mean the primary goal is to differentiate index scan plans with the same cost?
> 2. If so, how likely is for index scan that used to cost less than other scan types (sequential, bitmap) to cost more with this additional cost?

Through the PR we want to differentiate between the scenarios where earlier cost was same, as shared in the description. 
Also it is very unlikely that an index scan with lower cost will override the other scan types as the additional cost is introduced in the order of 10^(-4). So in a unlikely even of same cost for all the different types of scans this additional cost may prove to be decisive. 
 The 'explain pipeline' with this branch, showed no plan diff due to additional cost factor.

24nishant commented:
> The following two examples could use some rewording.
> 
> > Differentiate in index scan cost if for the same index, the query is on different
> > index columns, earlier this cost was same.
> > Eg: For Index - idx_foo_ab ; predicate - a vs predicate - b vs predicate - a,b
> 
> And just to be sure, this example illustrates the missing index cost?
> 
> > Differentiate in index scan cost if for the same index, different predicate
> > columns are missing in the queries, earlier this cost was same.
> > Eg: For Index - idx_foo_a ; predicate - a vs predicate - a,b vs predicate - a,b,c
> 
> and this example illustrates the missing pred cost?
> 
> And before this PR, if there's any overlap between the index and predicate columns, we cost the index scan the same amount?

1. I have reworded the examples 
2.  Can you please give an example for the overlap case?

24nishant commented:
> I am not persuaded that `m_ulResidualPredicateSize` belongs to the `IndexScan/Get` classes. I see the convenience obtaining its value in `CXformUtils`, where we derive the residual predicates expression array. But we should only add a class member variable if the variable speaks of what this class is, or, if we want the variable to be easily accessible. That's not the case with `m_ulResidualPredicateSize`. When we think of `IndexScan/Get`, I don't see the number of the predicates that we have to filter again on each row fetched a key signature of the class. And the only occasion that we access the variable is in the cost model. In my opinion, it makes a lot more sense to derive the number of residual predicates in `CCostModelGPDB`. Actually we can even make `ComputeAdditionalMIssingIndexWeight` dual-purposed, where we have access to both index and predicate columns. The first round we locate the indices of the index columns unused by predicates. The second round we locate the count of unindexed predicate columns.

Originally, I planned to do it the way you described above, I wanted to compute missing index/predicate columns in a function and send single value to the CostIndexScan().
I implemented in the current way, because as we reach to the costing stage, for index scans, from all the predicates that are present in the query we are left only with the predicates which are applicable on the index.  Thus, we won't be able to compute the 'missing predicate weight'
Thus,  I had to pass this information to the costing stage.  Since, it was required for all the 'index types' I added it to the class so that it is easily accessible.

THANATOSLAVA commented:
> as we reach to the costing stage, for index scans, from all the predicates that are present in the query we are left only with the predicates which are applicable on the index.

So take this example, table foo is indexed on `a,b`, and the query predicates are on columns `b`, `c`. Were you suggesting that in the cost model you only had access to column `b`, but not column `c`? Why would that be? The expression handle should allow us to access everything, even without the variable attached to the operator directly.

THANATOSLAVA commented:
> > as we reach to the costing stage, for index scans, from all the predicates that are present in the query we are left only with the predicates which are applicable on the index.
> 
> So take this example, table foo is indexed on `a,b`, and the query predicates are on columns `b`, `c`. Were you suggesting that in the cost model you only had access to column `b`, but not column `c`? Why would that be? The expression handle should allow us to access everything, even without the variable attached to the operator directly.

Hmmm researched a bit more and you're right it's tricky. Group expression `IndexScan`'s only child group is the index condition. Its sibling group is the filter. But depending on the parent group, sibling groups could be different, so it won't make sense to look for the sibling.

```
	// Find all the columns of index which are not present in the predicate.
	// Eg- index : a,b,c  Predicate: c,d; so Missing index column = a,b
```

In the comment here, we need to add "Used Predicate: c", and clarify that we locate `a` and `b` using the index columns and the used predicate columns.

THANATOSLAVA commented:
> 1. I have reworded the examples
> 2. Can you please give an example for the overlap case?

Much better. We can leave out the chart though. Take scenario 1 as an example, we just need the summary plus the queries.

1. Cost tuning to account for unused index columns in the predicates
```
create index idx_foo_a on foo using btree(a);
select * from foo where a = 1;
select * from foo where a = 1 and b = 1;
select * from foo where a = 1 and b = 1 and c = 1;
```

THANATOSLAVA commented:
> Also it is very unlikely that an index scan with lower cost will override the other scan types as the additional cost is introduced in the order of 10^(-4).

Then we need to be clear that this commit is meant to break a tie among different index scan alternatives. It's not meant to raise the index scan cost in general (which is what we are doing), so it's more likely other scan types with better performance can be selected.

> So in a unlikely even of same cost for all the different types of scans this additional cost may prove to be decisive.

This statement has to be backed by data. If we don't have data, we need to be careful about calling `10^-4` decisive. Cause the next question would be why not `10^-3` or `10^-5`. If both `10^-3` and `10^-5` could have worked, `10^-4` is apparently not decisive.

Actually it would be nice to add a statement like "The cost of sequential scan is typically two orders of magnitude (need verification) above the index scan, so it's unlikely the cost tuning in this commit results in the selection of a different scan type."

THANATOSLAVA commented on src/backend/gporca/libgpdbcost/src/CCostModelGPDB.cpp:
```diff
@@ -1630,21 +1728,62 @@ CCostModelGPDB::CostIndexScan(CMemoryPool *,  // mp
+	const COptCtxt *poctxt = COptCtxt::PoctxtFromTLS();
+	CMDAccessor *md_accessor = poctxt->Pmda();
+	CColRefArray *pdrgpcrIndexColumns = nullptr;
+	ULONG ulCummulativeMissingPredWeight = 0;
```
Is there weight? I thought it's just count?

THANATOSLAVA replied:
Then I would suggesting calling it "count" or "size" or "nums". Variable names should speak of what they are.

24nishant commented:
> > Also it is very unlikely that an index scan with lower cost will override the other scan types as the additional cost is introduced in the order of 10^(-4).
> 
> Then we need to be clear that this commit is meant to break a tie among different index scan alternatives. It's not meant to raise the index scan cost in general (which is what we are doing), so it's more likely other scan types with better performance can be selected.
> 
> > So in a unlikely even of same cost for all the different types of scans this additional cost may prove to be decisive.
> 
> This statement has to be backed by data. If we don't have data, we need to be careful about calling `10^-4` decisive. Cause the next question would be why not `10^-3` or `10^-5`. If both `10^-3` and `10^-5` could have worked, `10^-4` is apparently not decisive.
> 
> Actually it would be nice to add a statement like "The cost of sequential scan is typically two orders of magnitude (need verification) above the index scan, so it's unlikely the cost tuning in this commit results in the selection of a different scan type."


I tried few simple queries, to check the cost difference between Index scans, sequential scans and bitmap scans. 
While costing, Bitmap scans ([Ref1](https://github.com/greenplum-db/gpdb/blob/d77a45911c173496f0a7242774f3b93384447fe3/src/backend/gporca/libgpdbcost/src/CCostModelGPDB.cpp#L1889), [Ref2](https://github.com/greenplum-db/gpdb/blob/d77a45911c173496f0a7242774f3b93384447fe3/src/backend/gporca/libgpdbcost/src/CCostModelGPDB.cpp#L1999)) & sequential scans ([Ref](https://github.com/greenplum-db/gpdb/blob/d77a45911c173496f0a7242774f3b93384447fe3/src/backend/gporca/libgpdbcost/src/CCostModelGPDB.cpp#L2109)), **the cost equation has a fixed 'Initial cost' mapped 
to DInitScanFacorVal (431.0).** 

This factor is not present in the "Index scan costing" (Index scan [Ref](https://github.com/greenplum-db/gpdb/blob/d77a45911c173496f0a7242774f3b93384447fe3/src/backend/gporca/libgpdbcost/src/CCostModelGPDB.cpp#L1668), Index only scan [Ref](https://github.com/greenplum-db/gpdb/blob/d77a45911c173496f0a7242774f3b93384447fe3/src/backend/gporca/libgpdbcost/src/CCostModelGPDB.cpp#L1802)) . Thus, i believe it is safe to 
assume, that changing cost in the order of 10e-4, should not result in any change of table scan operators .

For a simple query, i checked the plans for these three scan types :

```
drop table if exists foo;
create table foo (a int , b int, c int) distributed by (a);
insert into foo select i,i,i from generate_series(1,5)i;
insert into foo select 1,1,1*i/i from generate_series(1,5)i;
create index idx_foo_ab ON foo USING btree(a,b);
create index idx_a_bitmap on foo_ using bitmap (a);
analyze foo;
```

**INDEX SCAN** 
```
nishants4=# explain analyze select * from foo where a=1;

 Gather Motion 1:1  (slice1; segments: 1)  (cost=0.00..6.00 rows=6 width=12) (actual time=0.614..0.615 rows=6 loops=1)
   ->  Index Scan using idx_foo_ab on foo  (cost=0.00..6.00 rows=2 width=12) (actual time=0.020..0.027 rows=6 loops=1)
         Index Cond: (a = 1)
 Optimizer: Pivotal Optimizer (GPORCA)
 Planning Time: 4.037 ms
   (slice0)    Executor memory: 39K bytes.
   (slice1)    Executor memory: 42K bytes (seg1).
 Memory used:  128000kB
 Execution Time: 1.037 ms
(9 rows)
```


**BITMAP SCAN** 
```
nishants4=# set optimizer_enable_indexscan to off;
nishants4=# explain analyze select * from foo where a=1;

 Gather Motion 1:1  (slice1; segments: 1)  (cost=0.00..397.97 rows=6 width=12) (actual time=1.143..1.144 rows=6 loops=1)
   ->  Bitmap Heap Scan on foo  (cost=0.00..397.97 rows=2 width=12) (actual time=0.246..0.270 rows=6 loops=1)
         Recheck Cond: (a = 1)
         ->  Bitmap Index Scan on idx_a_bitmap  (cost=0.00..0.00 rows=0 width=0) (actual time=0.188..0.188 rows=1 loops=1)
               Index Cond: (a = 1)
 Optimizer: Pivotal Optimizer (GPORCA)
 Planning Time: 3.712 ms
   (slice0)    Executor memory: 43K bytes.
   (slice1)    Executor memory: 636K bytes (seg1).
 Memory used:  128000kB
 Execution Time: 1.951 ms
(11 rows)
```



SEQUENTIAL SCAN 
```
nishants4=# set optimizer_enable_bitmapscan to off;
nishants4=# explain analyze select * from foo where a=1;

 Gather Motion 1:1  (slice1; segments: 1)  (cost=0.00..431.00 rows=6 width=12) (actual time=0.935..0.937 rows=6 loops=1)
   ->  Seq Scan on foo  (cost=0.00..431.00 rows=2 width=12) (actual time=0.449..0.453 rows=6 loops=1)
         Filter: (a = 1)
 Optimizer: Pivotal Optimizer (GPORCA)
 Planning Time: 4.434 ms
   (slice0)    Executor memory: 36K bytes.
   (slice1)    Executor memory: 36K bytes (seg1).
 Memory used:  128000kB
 Execution Time: 1.415 ms
(9 rows)
```

chrishajas commented:
Can you also run the explain pipeline with these changes? Also, why are costs in other MDPs that don't involve indexes changing?

dgkimura commented on src/backend/gporca/libgpdbcost/src/CCostModelGPDB.cpp:
```diff
@@ -1630,21 +1720,62 @@ CCostModelGPDB::CostIndexScan(CMemoryPool *,  // mp
 		indexIncludedArray = CPhysicalIndexScan::PopConvert(pop)
 								 ->Pindexdesc()
 								 ->PdrgpcoldescIncluded();
+		const IMDRelation *pmdrel = md_accessor->RetrieveRel(
+			CPhysicalIndexScan::PopConvert(pop)->Ptabdesc()->MDId());
+
+		const IMDIndex *pmdindex = md_accessor->RetrieveIndex(
+			CPhysicalIndexScan::PopConvert(pop)->Pindexdesc()->MDId());
+
+		pdrgpcrIndexColumns = CXformUtils::PdrgpcrIndexKeys(
+			mp, CPhysicalIndexScan::PopConvert(pop)->PdrgpcrOutput(), pmdindex,
+			pmdrel);
+
+		ulCummulativeMissingPredCount =
+			CPhysicalIndexScan::PopConvert(pop)->ResidualPredicateSize();
+
+		stats = CPhysicalIndexScan::PopConvert(pop)->PstatsBaseTable();
```
It may be worth pulling this duplicate code out into a templated function.  Or since we have C++14 you maybe able to do a generic lambda.  ;)

dgkimura commented on src/backend/gporca/libgpdbcost/src/CCostModelGPDB.cpp:
```diff
@@ -1724,6 +1876,15 @@ CCostModelGPDB::CostIndexOnlyScan(CMemoryPool *mp GPOS_UNUSED,	  // mp
 		indexIncludedArray = CPhysicalIndexOnlyScan::PopConvert(pop)
 								 ->Pindexdesc()
 								 ->PdrgpcoldescIncluded();
+		const IMDRelation *pmdrel = md_accessor->RetrieveRel(
+			CPhysicalIndexOnlyScan::PopConvert(pop)->Ptabdesc()->MDId());
+
+		const IMDIndex *pmdindex = md_accessor->RetrieveIndex(
+			CPhysicalIndexOnlyScan::PopConvert(pop)->Pindexdesc()->MDId());
+
+		pdrgpcrIndexColumns = CXformUtils::PdrgpcrIndexKeys(
+			mp, CPhysicalIndexOnlyScan::PopConvert(pop)->PdrgpcrOutput(),
+			pmdindex, pmdrel);
```
Same, consider generic lambda or template function.

24nishant commented:
> Can you also run the explain pipeline with these changes? Also, why are costs in other MDPs that don't involve indexes changing?


I ran the explain pipeline for these changes:

1. Plan diffs were found in the 'Explain pipeline' run for this PR branch.
2. These were due to the added logic of ' Enabling computation of statistics for columns used in the query for the table ([here](https://github.com/24nishant/gpdb/blob/0530da6242a3ab2d3c73c7e8e6290f9d2c3ef030/src/backend/gporca/libgpopt/src/operators/CPhysicalScan.cpp#L260))'
3. **However, there are no 'diffs' in the 'Explain pipeline' due to change in costing.**
This was ascertained, by creating a  [new](https://github.com/greenplum-db/gpdb/compare/main...24nishant:gpdb:Index_Costing_7X_NoBaseTabStats) branch (out of this PR branch) & removing the above described logic for statistics computation. On running, 'explain pipeline' for this new branch, no plan diffs were found.
 Here is a summary of MDP changes, for this branch. The detailed list of changes (MDP vs change), I have mentioned in [this](https://docs.google.com/spreadsheets/d/1aDKGUFSFi5XhpQI8XdugwKTgKWo6RjDL/edit?usp=sharing&ouid=105599807536520562474&rtpof=true&sd=true) excel sheet. 

<img width="299" alt="image" src="https://github.com/greenplum-db/gpdb/assets/99574850/eda9c0c2-6525-4a6b-b8ba-9f859bd11615">

24nishant commented on src/test/regress/expected/gp_index_costing_optimizer.out:
```diff
@@ -0,0 +1,233 @@
+set optimizer_enable_indexscan to off;
+CREATE INDEX idx_foo_abc ON foo USING btree(a,b,c);
+CREATE INDEX idx_foo_cba ON foo USING btree(c,b,a);
+vacuum analyze foo;
+    --  Index idx_foo_abc should be selected
+explain select * from foo where a=1;
+                                    QUERY PLAN
+-----------------------------------------------------------------------------------
+ Gather Motion 1:1  (slice1; segments: 1)  (cost=0.00..6.00 rows=6 width=12)
+   ->  Index Only Scan using idx_foo_abc on foo  (cost=0.00..6.00 rows=2 width=12)
+         Index Cond: (a = 1)
+ Optimizer: Pivotal Optimizer (GPORCA)
+(4 rows)
```
Thanks, @dgkimura for sharing this example. 
I ran dev pipeline multiple times, and had a consistent green build for this test and all of the other tests added in gp_index_costing.sql However, to be on the safer side, I can add this test serially in greenplum_schedule.

24nishant commented on src/backend/gporca/libgpdbcost/src/CCostModelGPDB.cpp:
```diff
@@ -1578,6 +1579,95 @@ CCostModelGPDB::CostSequenceProject(CMemoryPool *mp, CExpressionHandle &exprhdl,
+//
+//---------------------------------------------------------------------------
+CDouble
+CCostModelGPDB::ComputeMissingIndexWeight(CExpressionHandle &exprhdl,
```
Thanks @dgkimura

24nishant commented on src/backend/gporca/libgpdbcost/src/CCostModelGPDB.cpp:
```diff
@@ -1578,6 +1579,95 @@ CCostModelGPDB::CostSequenceProject(CMemoryPool *mp, CExpressionHandle &exprhdl,
+
+				CDouble dTableRows = CStatistics::CastStats(stats)->Rows();
+
+				GPOS_ASSERT(0 != dTableRows);
```
Since `dTableRows` is used as a divisor, so I checked it before division.
I dont expect it to be zero, just wanted to be sure before division.

24nishant replied:
1. We have an assertion in [constructor](https://github.com/greenplum-db/gpdb/blob/6a368f33f92579ca641589a4e43b34dcbe367400/src/backend/gporca/libnaucrates/src/statistics/CStatistics.cpp#L80) of CStatistics class, so a ZERO should not come in dTableRows. But since, it is used in the denominator, I would want to check it before using. 

2. [0/0] case can only happen if rows = 0, because of the following
Ratio = dNdv / dTableRows
NDV =min (rows, upperBoundNDV of Column)
dTableRows = rows

Now 
- if rows = 0, and we have put an assertion for it then, [0/0] won't happen.
- if row != 0, then we will get some value for NDV, and a valid value for the Ratio.

24nishant commented on src/backend/gporca/libgpdbcost/src/CCostModelGPDB.cpp:
```diff
@@ -1578,6 +1579,95 @@ CCostModelGPDB::CostSequenceProject(CMemoryPool *mp, CExpressionHandle &exprhdl,
+			CDouble dMissingIndexColWeight = 0;
+			bool fIndexPredColMatchFound = false;
+
+			// For every index column, we check if it is present in the
+			// predicate columns
+			CColRefSetIter iter(*pcrsUsedPredicate);
+			while (iter.Advance())
+			{
+				CColRef *colrefPredColm = iter.Pcr();
+				if (CColRef::Equals(colrefIndexColumn, colrefPredColm))
+				{
+					fIndexPredColMatchFound = true;
+					break;
+				}
+			}
```
Yes, I have updated the function using FMember().

24nishant commented on src/test/regress/expected/join_optimizer.out:
```diff
@@ -6652,13 +6652,14 @@ where j1.id1 % 1000 = 1 and j2.id1 % 1000 = 1;
    ->  Nested Loop
          Join Filter: true
-         ->  Seq Scan on j1
-               Filter: ((id1 % 1000) = 1)
-         ->  Index Scan using j2_pkey on j2
+         ->  Broadcast Motion 3:3  (slice2; segments: 3)
+               ->  Seq Scan on j1
+                     Filter: ((id1 % 1000) = 1)
+         ->  Index Only Scan using j2_pkey on j2
                Index Cond: ((id1 = j1.id1) AND (id2 = j1.id2))
                Filter: ((id1 % 1000) = 1)
  Optimizer: Pivotal Optimizer (GPORCA)
-(9 rows)
```
I have fixed this test. Now there is no plan diff.
The reason for this selection this plan was selection of Index Only scan. 

I am using a count of unused predicates on the index [here](https://github.com/greenplum-db/gpdb/blob/d168b06ebe043c610b5880143180be360a4e1498/src/backend/gporca/libgpopt/src/xforms/CXformUtils.cpp#L2487)
This value is computed in` ExtractIndexPredicates() `and is available in `pdrgpexprResidual`. 
This variable also holds predicate conditions not applicable on the index [here](https://github.com/greenplum-db/gpdb/blob/ba2cde2c2226e4503375caf54e8d4575ff51876a/src/backend/gporca/libgpopt/src/operators/CPredicateUtils.cpp#L2184) 
In the given query, this condition is satisfied which causes increase in number of unused predicate on index. This caused increase in cost for Index scan. 
Since this predicate is not applicable, I have now adjusted the count for it.

dgkimura commented on src/backend/gporca/libgpdbcost/src/CCostModelGPDB.cpp:
```diff
@@ -1580,6 +1581,129 @@ CCostModelGPDB::CostSequenceProject(CMemoryPool *mp, CExpressionHandle &exprhdl,
+		// Iterate through all the index columns to find, any unused
+		// index column in the predicate
+		// Eg- index : a,b,c  Predicate: c,d
+		// Used Predicate colm -'c'  (in the expHandle we have information of
```
s/colm/column

24nishant replied:
Thanks, Corrected in the code.

dgkimura commented on src/backend/gporca/libgpdbcost/src/CCostModelGPDB.cpp:
```diff
@@ -1580,6 +1581,129 @@ CCostModelGPDB::CostSequenceProject(CMemoryPool *mp, CExpressionHandle &exprhdl,
+//		CCostModelGPDB::GetCommonIndexData
+//
+//	@doc:
+// 		Get  count of index keys, width of included col, array of index columns
```
Extra space between "Get" and "count".

24nishant replied:
Thanks, Corrected in the code.

dgkimura reviewed (approved):
Looks good.  I think offline you mentioned creating a separate PR for NDV column stats collection portion of this change.  Is that still the case?

24nishant commented on src/backend/gporca/data/dxl/minidump/Gist-NonPart-Lossy-BitmapIndexPlan.mdp:
```diff
@@ -269,16 +269,16 @@ Optimizer status: PQO version 2.65.1
             <dxl:Comparison ComparisonOperator="&lt;@" OperatorMdid="0.489.1.0">
               <dxl:Ident ColId="1" ColName="p" TypeMdid="0.604.1.0"/>
-              <dxl:ConstValue TypeMdid="0.604.1.0" Value="AAAAaAQAAAAAAAAAAAAUQAAAAAAAABRAAAAAAAAAAAAAAAAAAAAAAAAAAAAA&#10;AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAABRAAAAAAAAAFEAAAAAAAAAUQAAA&#10;AAAAABRAAAAAAAAAAAA="/>
+              <dxl:ConstValue TypeMdid="0.604.1.0" Value="AAAAaAQAAAAAAAAAAAAUQAAAAAAAABRAAAAAAAAAAAAAAAAAAAAAAAAAAAAA&#xA;AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAABRAAAAAAAAAFEAAAAAAAAAUQAAA&#xA;AAAAABRAAAAAAAAAAAA="/>
```
I have reverted these changes and strangely no error came locally in Ctests now.

24nishant commented on src/backend/gporca/data/dxl/minidump/DynamicIndexOnlyScanCosting.mdp:
```diff
@@ -0,0 +1,685 @@
+        Without data the cost for both of them is same. This is because, the only difference is due to 'dPartialVisFrac'
+        while computing cost.
+
+         explain  select *   from bar_PT join foo on bar_PT.a =foo.a and bar_PT.c =foo.c;
```
Thanks, I have updated the MDP.

24nishant commented on src/backend/gporca/data/dxl/minidump/DynamicIndexScanCosting.mdp:
```diff
@@ -0,0 +1,685 @@
+         create table bar_PT (a int, b int, c int) partition by range(a) (start (1) inclusive end (12) every (2)) distributed by (a);
+         CREATE INDEX idx_bar_PT_ab ON bar_PT USING btree(a,b);
+         analyze bar_PT;
+         explain  select *   from bar_PT join foo on bar_PT.a =foo.a and bar_PT.c =foo.c;
```
Thanks, I have updated the MDP.

THANATOSLAVA commented on src/backend/gporca/libgpdbcost/src/CCostModelGPDB.cpp:
```diff
@@ -1628,29 +1752,32 @@ CCostModelGPDB::CostIndexScan(CMemoryPool *,  // mp
+	const COptCtxt *poctxt = COptCtxt::PoctxtFromTLS();
+	CMDAccessor *md_accessor = poctxt->Pmda();
+	CColRefArray *pdrgpcrIndexColumns = nullptr;
+	ULONG ulCummulativeUnindexedPredCount = 0;
```
"Cumulative" redundant. "Count" is always cumulative.

24nishant replied:
Yup, that make sense in our case. Corrected !

THANATOSLAVA commented:
`ptr` --> `op`

THANATOSLAVA commented on src/backend/gporca/libgpopt/src/xforms/CXformUtils.cpp:
```diff
@@ -2473,13 +2473,42 @@ CXformUtils::PexprBuildBtreeIndexPlan(
+	// unused(not present in index) predicate size is taken as zero.
+	// 2. All the unsupported predicates on the index are also removed
+	// eg
+	//       create table j1 (a1 int, b1 int, primary key(a1,b1));
+	//      create table j2 (a2 int, b2 int, primary key(a2,b2));
+
+	// 		explain (costs off) select * from j1
+	//      inner join j2 on j1.a1 = j2.a2 and j1.b1 = j2.b2
+	//      where j1.a1 % 1000 = 1 and j2.a2 % 1000 = 1;
+	// here both the conditions in the 'where' clause are - 'not supported index
+	// predicate'.
```
This is a very good example.

Let's remove `j2.a2 % 1000 = 1` cause it's made redundant by `j1.a1 = j2.a2` and `j1.a1 % 1000 = 1`

We should also add:

1. `j1.a1 = j2.a2` and `j1.b1 = j2.b2` are the two predicates added to indexable predicates array.
2. `j1.a1 % 1000 = 1` refers to index key a1, but since the predicate isn't indexable, it's added to the residual predicates array. Correspondingly, our tally of unsupported predicates becomes 1.

24nishant replied:
I think we should keep `j2.a2 % 1000 = 1`, If we remove  it then for table J2 unsupported predicate will  become 0.
Thus only for table J1 we will get unsupported predicate.

THANATOSLAVA replied:
By the time when we sort index/residual predicates, ORCA would have already tossed `j2.a2 % 1000 = 1` out of the window. I don't think `ExtractIndexPredicates` ever processes `j2.a2 % 1000 = 1`.

You can make it `j2.a2 % 1000 = 2`. Then I think `ExtractIndexPredicates` would process all the four predicates.

24nishant replied:
I checked the plan for this query, it looks like ORCA does uses both the Modulus conditions in the original query. To make the example unambiguous, I have updated the comment with` j2.a2 % 1000 = 2`

`nishants4=# explain select * from j1 inner join j2 on j1.a1 = j2.a2 and j1.b1 = j2.b2  where j1.a1 % 1000 = 1 and j2.a2 % 1000 = 1;`
```
                                  QUERY PLAN
-------------------------------------------------------------------------------
 Gather Motion 3:1  (slice1; segments: 3)  (cost=0.00..437.00 rows=1 width=16)
   ->  Nested Loop  (cost=0.00..437.00 rows=1 width=16)
         Join Filter: true
         ->  Seq Scan on j1  (cost=0.00..431.00 rows=1 width=8)
               Filter: ((a1 % 1000) = 1)
         ->  Index Scan using j2_pkey on j2  (cost=0.00..6.00 rows=1 width=8)
               Index Cond: ((a2 = j1.a1) AND (b2 = j1.b1))
               Filter: ((a2 % 1000) = 1)
 Optimizer: GPORCA
(9 rows)
```

THANATOSLAVA commented on src/backend/gporca/libgpopt/src/operators/CPredicateUtils.cpp:
```diff
@@ -2182,6 +2182,7 @@ CPredicateUtils::ExtractIndexPredicates(
 			{
 				// not a supported predicate
 				pdrgpexprTarget = pdrgpexprResidual;
+				ulUnsupportedPredCount = ulUnsupportedPredCount + 1;
```
Given that you deduct `ulUnsupportedPredCount` from `pdrgpexprResidual->Size()`, why don't you do the tally in 

```c++
		if (!fSubset)
		{
			pdrgpexprResidual->Append(pexprCond);
			// keep tally of the unindexed predicates
			continue;
		}
```

24nishant replied:
I am sorry, I didn't get you. 
If you meant, why am I subtracting the count outside ?
Then, we get 'ulUnsupportedPredCount' after some additional check on the predicate conditions, and that value is not available here ( if (!fSubset)).  So I did it outside the function.

THANATOSLAVA replied:
Maybe I misinterpreted what this function does, but my understanding is

We append two types of predicates to `pdrgpexprResidual`: (1) predicates that don't refer to index keys at all, and (2) predicates that refer to index keys but don't qualify for index scan.

And you are looking for the number of type(1) predicates?

Right now what you do is keeping the tally of type(2) predicates. Later, you find out the number of type(1) predicates by deducting the type(2) count from the total.

What I was suggesting was, each time `fSubset == false`, a type(1) predicate gets appended to `pdrgpexprResidual`. So we can do the tally here, which saves you the deduction step.

THANATOSLAVA commented on src/backend/gporca/libgpopt/src/xforms/CXformUtils.cpp:
```diff
@@ -2473,13 +2473,42 @@ CXformUtils::PexprBuildBtreeIndexPlan(
+	// an Index scan is created (Eg: Order by some index columns), then this count
+	// is not valid for costing.
+	// eg: explain select a from foo order by a limit 10; {idx_a on table},
+	// pdrgpexprResidual->size() is > 0, but (pdrgpexprIndex->Size()=0)
```
We can be more precise:

```
pdrgpexprResidual->size() = 1
```

24nishant replied:
Updated!

THANATOSLAVA commented:
Also don't forget about fixing the wording of "missing index/predicate column" in your commit message

THANATOSLAVA commented on src/test/regress/expected/gp_index_costing_optimizer.out:
```diff
@@ -0,0 +1,233 @@
+-- 2.1 Test case for index scans
+CREATE INDEX idx_foo_a ON foo USING btree(a);
+analyze foo;
+-- The index scan cost for 'Query2' where two of the predicate columns (b,c) are missing
+-- in the index should be higher compared to Query1
+-- Query1
+explain select * from foo where a=1;
```
This experiment isn't the most effective mainly because in displaying the cost we only show two digits after the decimal point. I wonder if this sequence helps illustrate the point --

1. Create idx_a
2. Run query `a=1 and b=1 and c=1`. We should see index scan with a filter on `b=1 and c=1`
3. Create idx_abc
4. Run the same query `a=1 and b=1 and c=1`. We should see index scan using `idx_abc`

We can do the similar for dynamic index scan

24nishant commented on src/backend/gporca/libgpdbcost/src/CCostModelGPDB.cpp:
```diff
@@ -1580,6 +1581,129 @@ CCostModelGPDB::CostSequenceProject(CMemoryPool *mp, CExpressionHandle &exprhdl,
+	CDouble dCummulativeUnusedIndexWeight = 0;
+	CDouble dNdv(1.0);
+
+	// Finding used predicate columns
```
Please correct me if am wrong, 'pcrsUsedPredicate' holds all the columns used in the predicate.
Here, we have all the columns which are applicable on the index, no other column is present here. 
So should we simply say, 'Finding columns used in the predicate' ?

THANATOSLAVA replied:
You can't be wrong cause it's your variable ;)

I thought in your example `index : a,b,c  Predicate: c,d`, `pcrsUsedPredicate` is `c`?

If so, `pcrsUsedPredicate` should refer to all the index columns used in the predicates, no?

24nishant commented:
> Member

Yes @dgkimura , I am working on the PR.
I have created a [local](https://github.com/greenplum-db/gpdb/compare/main...24nishant:gpdb:TableStatsUpdate_7x) branch for it, the 'dev' pipeline is also green.
Will raise the PR next for review, working on some more ICW test cases.

24nishant commented on src/backend/gporca/libgpopt/src/operators/CPhysicalScan.cpp:
```diff
@@ -256,6 +256,18 @@ CPhysicalScan::ComputeTableStats(CMemoryPool *mp)
+			pcrsHist->Include(colref);
+		}
+	}
+
```
As discussed offline, I have now removed this change and I am working on raising a separate PR with this logic of 'table statistics computation'.

24nishant commented on src/backend/gporca/libgpdbcost/src/CCostModelGPDB.cpp:
```diff
@@ -1580,6 +1581,129 @@ CCostModelGPDB::CostSequenceProject(CMemoryPool *mp, CExpressionHandle &exprhdl,
+		// Iterate through all the index columns to find, any unused
+		// index columns in the predicate
+		// Eg- index : a,b,c  Predicate: c,d
+		// Used Predicate column -'c'  (in the expHandle we have information of
+		// only those predicate columns, which are applicable on the index)
```
Thanks, corrected in code.

24nishant commented on src/backend/gporca/libgpdbcost/src/CCostModelGPDB.cpp:
```diff
@@ -1664,11 +1791,28 @@ CCostModelGPDB::CostIndexScan(CMemoryPool *,  // mp
+		ComputeUnusedIndexWeight(exprhdl, pdrgpcrIndexColumns, stats) *
+		dIndexCostConversionFactor;
+
+	CDouble dUnindexedPredUnusedIndexCost =
+		dUnindexedPredCost + dUnusedIndexCost;
```
I have changed it.

24nishant commented on src/backend/gporca/libgpdbcost/src/CCostModelParamsGPDB.cpp:
```diff
@@ -191,6 +191,13 @@ const CDouble CCostModelParamsGPDB::DPenalizeHJSkewUpperLimit(10.0);
 // default scalar func cost
 const CDouble CCostModelParamsGPDB::DScalarFuncCost(1.0e-04);
 
+// Additional cost conversion factor for index & Index only scans.
+// Kept it as '1.0e-04' because using the additional cost factor,
+// primary aim is to differentiate between 'plans' with same cost.
+// Thus, a small value is sufficient to differentiate between them.
+// For eg cost for index idx_ab & idx_ba, will be same without the
+// additional cost component.
```
Yes, I have updated in the code.

24nishant commented on src/backend/gporca/libgpopt/src/xforms/CXformUtils.cpp:
```diff
@@ -2473,13 +2473,42 @@ CXformUtils::PexprBuildBtreeIndexPlan(
+	// here both the conditions in the 'where' clause are - 'not supported index
+	// predicate'.
+
+	ULONG ulResidualPredicateSize = 0;
+	if (pdrgpexprIndex->Size() > 0)
+	{
+		ulResidualPredicateSize =
+			pdrgpexprResidual->Size() - ulUnsupportedPredCount;
```
Yes, thanks,I agree.
To make it more clear, I have renamed the variable to 'ulUnindexedPredColCount' throughout the code.

24nishant commented on src/test/regress/sql/gp_index_costing.sql:
```diff
@@ -0,0 +1,116 @@
+
+drop table if exists foo;
+create table foo (a int , b int, c int) distributed by (a);
+insert into foo select i,i,i from generate_series(1,5)i;
+insert into foo select 1,1,1*i/i from generate_series(1,5)i;
```
Yes, I have added a comment. 
I added these duplicate values so that we have different values for NDV and table rows while costing. 
This way we can show the impact of NDV/table rows while costing

24nishant commented on src/test/regress/expected/dpe_optimizer.out:
```diff
@@ -270,23 +269,30 @@ select * from pt where ptid in (select tid from t where t1 = 'hello' || tid);
-               Buckets: 524288  Batches: 1  Memory Usage: 4097kB
-               ->  Partition Selector (selector id: $0) (actual rows=2 loops=1)
-                     ->  Broadcast Motion 3:3  (slice2; segments: 3) (actual rows=2 loops=1)
+   ->  Hash Join (actual rows=1 loops=1)
+         Hash Cond: (t.tid = pt.ptid)
+         Extra Text: (seg1)   Hash chain length 1.0 avg, 1 max, using 1 of 262144 buckets.
+         ->  GroupAggregate (actual rows=2 loops=1)
+               Group Key: t.tid
+               ->  Sort (actual rows=2 loops=1)
+                     Sort Key: t.tid
+                     Sort Method:  quicksort  Memory: 25kB
+                     Executor Memory: 26kB  Segments: 1  Max: 26kB (segment 1)
+                     ->  Redistribute Motion 3:3  (slice2; segments: 3) (actual rows=2 loops=1)
+                           Hash Key: t.tid
                            ->  Seq Scan on t (actual rows=2 loops=1)
                                  Filter: (t1 = ('hello'::text || (tid)::text))
+         ->  Hash (actual rows=1 loops=1)
+               Buckets: 262144  Batches: 1  Memory Usage: 2049kB
+               ->  Redistribute Motion 3:3  (slice3; segments: 3) (actual rows=1 loops=1)
+                     Hash Key: pt.ptid
+                     ->  Dynamic Index Scan on pt1_idx on pt (actual rows=1 loops=1)
+                           Index Cond: (pt1 = 'hello1'::text)
+                           Number of partitions to scan: 6 (out of 6)
+                           Partitions scanned:  Avg 6.0 x 3 workers.  Max 6 parts (seg0).
  Optimizer: Pivotal Optimizer (GPORCA)
 (16 rows)
```
This happened because now we don't have correct NDV as the table stats logic was removed.

- **With Correct NDV:** 
 In costing function NDV- 6 , total rows - 54
Cost of Index only scan using ptid_pt1_idx - 6.000187
Cost of Index scan using ptid_pt1_idx         - 6.000294
Cost of Index scan using pt1_idx                  - 6.000194

Thus index only scan using ptid_pt1_idx was selected earlier.

- **With In-Correct NDV (After removing stats logic)**
In costing function NDV- 54 , total rows - 54
Cost of Index only scan using ptid_pt1_idx - 6.000365
Cost of Index scan using ptid_pt1_idx         - 6.000477
Cost of Index scan using pt1_idx                  - 6.000194

Thus index scan using pt1_idx was selected earlier, which triggered overall a different plan.

THANATOSLAVA replied:
? But there appears to be more than the change in the index choice. The query is under the dpe test and we used to get DPE. Now it's not running DPE.

Plus, the plan counts 21 rows, instead of 16 rows?

24nishant replied:
I checked in the memo, from the ROOT group in this case 

- CPhysicalInnerHashJoin (High)[ 63 16 17 ]  **is selected.** while CPhysicalLeftSemiHashJoin (High) [ 16 8 17 ] is not due to high cost of Semi join.
- The cost increased for the Semi Join due to incorrect NDV. 
- Since both have different Childs, we have completely different plan.

THANATOSLAVA replied:
As we discussed:

1. Add data to the partition table so DPE is more likely to excel. We should look into the commit messages and understand the intent of the test. We need to justify our changes to existing tests.
2. If we have math proof that different scan types have no overlap in their cost ranges, we should add that to the commit message. That justifies our choice of "cost conversion factor", especially now that you are considering changing it from 10^-4 to 10^-6 to avoid this plan diff.
3. I still think limiting the stats collection to index scans can solve a lot of problems (MDP diffs, plan diffs). If 1 or 2 doesn't, definitely give this one a shot.

24nishant replied:
I tried the query with more data in the table.
The original query had 54 rows in the table, I changes it to 2500, 10000 and check the plan and performance. 

<img width="438" alt="image" src="https://github.com/greenplum-db/gpdb/assets/99574850/554b71eb-b076-48d3-b327-0e9dd6aab536">



For rows>2500, we get no plan difference. 

I also checked the intent of the test, it is to show the dynamic index only scan capabilities.

THANATOSLAVA replied:
My interpretation for the comparison above is:

1. The execution time is in the order of 20ms, so the following statements might just be within the error bars
2. There's a minor cost associated with correcting the NDV (19.3ms > 17.9ms)
3. Even with the additional cost due to the NDV correction, given the original query with 54 rows, the DPE plan is still superior to the wo/DPE plan (18.5ms < 19.4ms)

The bottom line is, since this test was added to the dpe suite, we need to ensure the plan uses DPE. As far as the actual query still returns 1 row, we can bump up the input size for now. This isn't ideal, because we are only doing so to hide an issue (obsolete NDV) exposed by this PR. We should revisit this test when the NDV stats is corrected in the next PR, and hopefully reduce the input size back to 54 rows.

THANATOSLAVA commented on src/backend/gporca/libgpopt/src/xforms/CXformUtils.cpp:
```diff
@@ -2472,14 +2472,35 @@ CXformUtils::PexprBuildBtreeIndexPlan(
 		CUtils::PcrsExtractColumns(mp, pdrgpexprIndex);
 	outer_refs_in_index_get->Intersection(outer_refs);
 
+	// In a case, where there is no predicate on the index columns but still
+	// an Index scan is created, (Eg: Order by some index columns)
+	// then ulUnindexedPredColCount is not valid for costing.
+	// eg: explain select a from foo order by a limit 10; {idx_a on table},
+	// ulUnindexedPredColCount = 1, but (pdrgpexprIndex->Size()=0)
+	if (pdrgpexprIndex->Size() == 0)
+	{
+		ulUnindexedPredColCount = 0;
+	}
+
```
I think it's better to move this logic to `CPredicateUtils::ExtractIndexPredicates` since `ulUnindexedPredColCount` isn't used anywhere besides the cost model.

24nishant replied:
Yes, I have updated it.

THANATOSLAVA commented on src/backend/gporca/libgpdbcost/include/gpdbcost/CCostModelGPDB.h:
```diff
@@ -201,6 +201,23 @@ class CCostModelGPDB : public ICostModel
 								   const CCostModelGPDB *pcmgpdb);
 
+
+	// Calculating additional columns weight for mismatch in index columns
+	// wrt predicate columns, for an Index Scan & Index Only Scan.
```
-->

"Calculating the weightage of index columns not used in predicates. For index scan & index only scan."

24nishant replied:
Updated.

THANATOSLAVA commented on src/backend/gporca/libgpdbcost/src/CCostModelGPDB.cpp:
```diff
@@ -1580,6 +1581,130 @@ CCostModelGPDB::CostSequenceProject(CMemoryPool *mp, CExpressionHandle &exprhdl,
+	return dCummulativeUnusedIndexWeight;
+}
+
+//---------------------------------------------------------------------------
+//	@function:
+//		CCostModelGPDB::GetCommonIndexData
+//
+//	@doc:
+// 		Get count of index keys, width of included col, array of index columns
+// 		and table statistics for Index Scan, Index only scan, Dynamic Index scan
+// 		& Dynamic Index only scan
```
Let's specify here it's the caller function's responsibility to release the memory for `pdrgpcrIndexColumns` and `stats`? Need to verify

24nishant replied:
Yes, verified and updated.

THANATOSLAVA commented on src/test/regress/expected/join_optimizer.out:
```diff
@@ -6645,7 +6645,7 @@ where j1.id1 % 1000 = 1 and j2.id1 % 1000 = 1;
                Filter: ((id1 % 1000) = 1)
  Optimizer: Pivotal Optimizer (GPORCA)
-(9 rows)
+(10 rows)
```
Why did the row count change when there's no plan change?

24nishant replied:
During the PR it changed and then was reverted back. 
The number of rows, were left unchanged. I have updated them also.

THANATOSLAVA commented:
Awesome work! Thanks for patiently addressing all my questions.

24nishant commented:
Thank you @dgkimura , @THANATOSLAVA  and @chrishajas for giving so many constructive feedback and guiding me throughout the PR.
